### PR TITLE
Easy Dropbear support for full disk encryption at rest.

### PR DESCRIPTION
**Pull request by gnat**
Based on recent full disk encryption support (CRYPT setting) added in https://github.com/hetzneronline/installimage/commit/5589fd2a2eda5a4cae87bf63707e641d6a09e5b6

Effort to help solve https://github.com/hetzneronline/installimage/issues/37 with additions from https://github.com/hetzneronline/installimage/pull/21 to provide remote SSH unlock of encryption using Dropbear.

* Hetzner Root and Cloud servers work.
* Ubuntu, Debian supported.
* Function stubs ready for Arch, Centos, Coreos, Suse.

Other contributors can fill out the stub for their alternative OS. I do not use Arch, etc. on servers, so it is out of scope of my contribution.

## Quickstart for Ubuntu 20.04

Set a public SSH key in Key Management area for your Rescue system before you begin. This is the key Dropbear will use to authenticate you.

#### SSH into server.
```bash
git clone --single-branch --branch crypt_dropbear https://github.com/gnat/installimage.git
alias installimage='/root/installimage/installimage'
nano autosetup
```

#### Content of autosetup.
```bash
# Encryption.
CRYPT 1
CRYPTPASSWORD test-password
CRYPTDROPBEAR 1
# Optionally specify CRYPTDROPBEARPORT
# Drives.
DRIVE1 /dev/nvme0n1
DRIVE2 /dev/nvme1n1
# RAID mirror.
SWRAID 1
SWRAIDLEVEL 1
# Other.
BOOTLOADER grub
HOSTNAME test-host
# Partitions.
PART /boot ext3 2G
PART lvm   vg0  all crypt
# Logical Volumes.
LV vg0 tmp  /tmp      ext4  16G
LV vg0 log  /var/log  ext4  16G
LV vg0 swap swap      swap  32G
LV vg0 root /         ext4  all
# Install image.
IMAGE /root/.oldroot/nfs/install/../images/Ubuntu-2004-focal-64-minimal.tar.gz
```

#### Begin installation.
```bash
installimage -c /root/autosetup -f yes -a
```

#### After reboot run cryptroot-unlock to unlock your disk.

Enjoy and please merge! This contribution is public domain.

### CODE REVIEW

**Comment by asciiprod** [comment]
Thank you very much for your contribution. In order to include your PR please add support for all supported distros and the necessary checks to ensure that a dropbear-supported key is present. The initramfs drop-in for RFC3442 routes required for Hetzner Cloud is also missing checks.

Due to the many constraints outlined above, it is unlikely we will add generic support.
For this reason it should be done via post-install as already mentioned in #21. 
Basically put the `install_crypt_dropbear`  function in a separate file and pass it to installimage with the `-x` option.

**Comment by gnat** [comment]
Thanks, but could we please get Hetzner management to allocate some resources into this? @asciiprod 

Considering the amount of attention this issue continues to receive in this repository, and how important encryption at rest is becoming (standard at AWS and Google Cloud), and how worse solutions to this exact issue for Hetzner tend to pop up online all the time:

* https://www.sysorchestra.com/hetzner-sx62-dedicated-server-with-software-raid-5-and/
* https://github.com/TheReal1604/disk-encryption-hetzner
* https://jkraemer.net/2018/04/fully-encrypted-headless-debian-stretch-setup
* https://docs.j7k6.org/headless-luks-encrypted-ubuntu-server/

This addition will be a necessary value-add for Hetzner in the future, and in-house resources really should be allocated to making setup and remote unlock of full disk encryption easy.

Keep in mind I'm already a paying customer and I don't really have time to do work on this, especially if it is to just be rejected in the end. There's a total lack of documentation on how post-install scripts work or anything else in installimage, makes it difficult for single contributors such as myself to fill your requirements without assistance.

**Comment by asciiprod** [comment]
Thanks for the feedback. You can basically pass any executable or script as post-install. We can look into adding a sample script.
If you think this topic is popular and should be made more public, consider submitting a [tutorial](https://community.hetzner.com/) 
for a 50 Euro credit to your account.

Please note that a provider-managed data at rest protection is not the same as a self-managed solution with LUKS.

**Comment by gnat** [comment]
Thanks but as mentioned I do not have a lot of extra time to put into this as an individual paying customer of Hetzner. Official contributions from Hetzner staff would be highly appreciated by everyone.

**Comment by laszlocph** [comment]
For anyone who finds this issue: running post-install scripts can be triggered with the `-x` switch like: `installimage -a -c /tmp/setup.conf -x /tmp/post-install.sh`

A working disk encryption tutorial can be found here: https://community.hetzner.com/tutorials/install-ubuntu-2004-with-full-disk-encryption

Just tested on Ubuntu 2204 on an EX43.
